spec/models/reservation_spec.rb

```diff
@@ -1,725 +1,913 @@
-    #   reservation.should be_not_in_past
-    #   reservation.should be_matched_item_and_model
-    #   reservation.should be_duration_allowed # fails: tries to run
-    # validations on nil
-    #   reservation.should be_start_date_is_not_blackout
-    #   reservation.should be_due_date_is_not_blackout
-    #   reservation.should be_available #fails: tries to run validations on nil
-    #   reservation.should be_quantity_eq_model_allowed # fails: tries to run
-    # validations on nil
-    #   reservation.should be_quantity_cat_allowed # fails: tries to run
-    # validations on nil
-    # end
-    it 'updates with equipment model' do
-      reservation.equipment_model = FactoryGirl.create(:equipment_model)
-      FactoryGirl.create(:equipment_item,
-                         equipment_model: reservation.equipment_model)
-      expect(reservation.save).to be_truthy
-      expect(reservation).to be_valid
-      expect(Reservation.all.size).to eq(1)
+    it 'returns true when the set has no duplicate items' do
+      set = Array.new(2) { |i| ReservationMock.new(equipment_item_id: i) }
+      expect(described_class.unique_equipment_items?(set)).to be_truthy
     end
   end
 
-  context 'editing the due date' do
-    context 'while not checked out' do
-      shared_examples_for 'does not affect the overdue status' do |res_type|
-        subject(:reservation) do
-          r = FactoryGirl.build(res_type)
-          r.save(validate: false)
-          r
-        end
-
-        it 'regardless of date' do
-          reservation.due_date = Time.zone.today - 1.day
-          expect { reservation.save }.not_to change { reservation.overdue }
-          reservation.due_date = Time.zone.today + 1.day
-          expect { reservation.save }.not_to change { reservation.overdue }
-        end
+  describe 'custom validations' do
+    it 'is invalid when start date after due date' do
+      res = FactoryGirl.build(:valid_reservation,
+                              start_date: Time.zone.today + 3.days,
+                              due_date: Time.zone.today)
+      expect(res.start_date_before_due_date).not_to be_empty
+      expect(res.valid?).to be_falsey
+    end
+    it 'is invalid when it has an item of a different model' do
+      item = FactoryGirl.create(:equipment_item)
+      model = FactoryGirl.create(:equipment_model)
+      res = FactoryGirl.build(:valid_reservation,
+                              equipment_model: model,
+                              equipment_item: item)
+      expect(res.matched_item_and_model).not_to be_empty
+      expect(res.valid?).to be_falsey
+    end
+    describe 'availability validation' do
+      it 'always passes checked in reservations' do
+        model = FactoryGirl.create(:equipment_model)
+        res = FactoryGirl.build(:checked_in_reservation,
+                                equipment_model: model)
+        allow(model).to receive(:num_available).and_return(0)
+        expect(res.available).to be_nil
```

Comment:
Returning true/false from the validation methods would definitely be an improvement! I don't think it will mess up the validation -- the only complication I can think of at the moment is that there might be places where the `errors` array is used.

Making these methods private is a good idea. I'll go ahead and open up an issue for this after I finish going through this review.